File: app/src/main/java/com/reactive/owm/domain/database/DatabaseGateway.java
package com.reactive.owm.domain.database;

import android.arch.persistence.room.Database;
import android.arch.persistence.room.RoomDatabase;
import android.arch.persistence.room.TypeConverters;

import com.reactive.owm.entities.City;
import com.reactive.owm.entities.FavoriteCityId;

import io.reactivex.Flowable;
import io.reactivex.functions.Function;

@Database(
        entities = {City.class, FavoriteCityId.class},
        version = 1,
        exportSchema = false)
@TypeConverters(CoordinatesConverter.class)
public abstract class DatabaseGateway extends RoomDatabase {

    public abstract CitiesTable getCitiesTable();

    public abstract FavoriteCityIdsTable getFavoriteCityIdsTable();

}

File: kotlin-app/src/main/java/com/ahmedismail/app/domain/database/DatabaseGateway.kt
package com.ahmedismail.app.domain.database

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.room.TypeConverters
import com.ahmedismail.app.entities.City
import com.ahmedismail.app.entities.FavoriteCityId
import kotlinx.coroutines.experimental.CommonPool
import kotlinx.coroutines.experimental.withContext
import java.io.File
import java.io.FileOutputStream
import java.io.InputStream

private const val DATABASE_NAME = "DatabaseGateway.db"

@Database(entities = [(City::class), (FavoriteCityId::class)], version = 1, exportSchema = false)
@TypeConverters(CoordinatesConverter::class)
abstract class DatabaseGateway : RoomDatabase() {

    abstract val citiesTable: CitiesTable

    abstract val favoriteCityIdsTable: FavoriteCityIdsTable

}

suspend fun database(context: Context) =
        copyDatabaseFromAssets(context)
                .let { Room.databaseBuilder(context, DatabaseGateway::class.java, DATABASE_NAME) }
                .build()


private suspend fun copyDatabaseFromAssets(context: Context) =
        context.getDatabasePath(DATABASE_NAME)
                ?.takeUnless { it.exists() }
                ?.let { copy(context, it) }


private suspend fun copy(context: Context, databaseFile: File) {
    withContext(CommonPool) {
        databaseFile.parentFile.mkdirs()
        context.assets.open(DATABASE_NAME)
                .use { copyByteArray(databaseFile, it) }
    }
}


private fun copyByteArray(databaseFile: File, assetsInputStream: InputStream) {
    FileOutputStream(databaseFile)
            .use { it.write(byteArray(assetsInputStream)) }
}

private fun byteArray(assetsFileInputStream: InputStream) =
        ByteArray(assetsFileInputStream.available())
                .also { assetsFileInputStream.read(it) }




